feat: [rfc]Add user trace cli for mme logs

wallyrb opened the pull request:
Signed-off-by: Wally Rodriguez B <wallyrb@fb.com>

## Summary

User trace cli should facilitate filtering user control sessions in mme logs. The cli has 3 functionalities for a given IMSI and mme log.

- 'list_imsi'
Provide a list of imsi and number of occurrences found in logs. (No input required)

- 'list_ue_id'
List mme and enb ue id pairs for a given IMSI. (IMSI input required)

- 'session_trace'
Output session trace for given user id(mme_id) and IMSI (IMSI and mme_id required). If is a known issue, it will provide the error and suggestions. Two known errors have been added(APN error and IE not supported).
I use regex to match different ue_id patterns.

Additional inputs

-p : "Path to the data directory"
-i: "imsi"
-m: "mme_id"

In general, we can use below flow for this cli.

1. Grab one imsi from list_imsi
2. From the IMSI selected get the ue_id found from list_ue_id
3. Output with session_trace using imsi + mme_id found in Step 1 and 2.


Examples: 
```
python3 user_trace_cli.py -p mme_apn.log list_ue_id -i 000000000000110 

IMSI: 000000000000110

mme_id          enodeb_id        
-------------------------
25 0x19         728 0x2d8            
26 0x1a         729 0x2d9            
27 0x1b         730 0x2da            
28 0x1c         731 0x2db            
29 0x1d         732 0x2dc   
```

```
python3 user_trace_cli.py -p mme_ie.log list_imsi                      

IMSI            Occurrences     
------------------------------
000000000000110 331            
000000000000111 2              
000000000000112 2              
000000000000113 2              
000000000000114 2              
000000000000115 2              
000000000000116 2              
000000000000117 2              
000000000000118 2              
000000000000119 2      
```

For session trace logs, please find in https://www.internalfb.com/phabricator/paste/view/P461580010

## Test Plan

list_ue_id: https://www.internalfb.com/phabricator/paste/view/P461580714
list_imsi: https://www.internalfb.com/phabricator/paste/view/P461580340
session_trace: https://www.internalfb.com/phabricator/paste/view/P461580010

## Additional Information

Documentation added

![image](https://user-images.githubusercontent.com/37117037/141268004-76071f51-2c3b-4e2b-a312-e1912389d213.png)

wallyrb committed:
feat: mme user trace cli v8

Signed-off-by: Wally Rodriguez B <wallyrb@fb.com>

pshelar commented:
can we read state from redis rather than from mme logs?

wallyrb commented:
> can we read state from redis rather than from mme logs?

@pshelar  thanks for checking. I believe mme logs provide more detail information for troubleshooting control plane user issues. I think Redis will provide a snapshot of the last state of the user. The purpose of cli is to provide the journey of control messaging for a specific user, and if there was a failure, what messages were found in the middle.  

For example session_trace:  https://www.internalfb.com/phabricator/paste/view/P461580010

The option "list_ue_id" will provide historically which ue id were found for this user in the logs.(how many "sessions"). Not sure if redis provides historical view, I believe only the last state(last mme, enodeb id)

- For "list_imsi", I think this info we can get it from redis. I added as utility to get which IMSI were found in the log.

So in general, we can use this flow for this cli.

1. Grab one imsi with list_imsi
2. Get the ue_id found from list_ue_id
3. Output with session_trace using imsi + mme_id found in Step 1 and 2.

(We can repeat step 3, with other mme_id to find previous sessions)

Also, we can use this as offline tool :) (without accessing to the AGW). Updated github with some of these comments for clarification.

ardzoht commented on lte/gateway/python/scripts/user_trace_cli.py:
```diff
@@ -0,0 +1,291 @@
+        parser = self.create_parser()
+        self.args = parser.parse_args()
+
+        f = open(self.args.path, "r")
```
```suggestion
        f = open(self.args.path, "r")
```
```suggestion
        with open(self.args.path, "r") as f:
```

ardzoht commented on lte/gateway/python/scripts/user_trace_cli.py:
```diff
@@ -0,0 +1,291 @@
+        Returns: Output number of unique IMSI found in logs
+        List imsi and number of ocurrences in logs
+        """
+        imsi_pattern = "\\b\\d{15}\\b"
```
```suggestion
        imsi_pattern = r"\b\d{15}\b"
```

ardzoht commented on lte/gateway/python/scripts/user_trace_cli.py:
```diff
@@ -0,0 +1,291 @@
+            if imsi in imsi_list:
+                imsi_list[imsi] += 1
+            else:
+                imsi_list[imsi] = 1
```
```suggestion
        imsi_list = defaultdict(int)
        lines = self.lines
        for line in lines:
            match = re.findall(imsi_pattern, line)
            if not match:
                continue
            imsi = match[0]
            imsi_list[imsi] += 1
```

ardzoht commented on lte/gateway/python/scripts/user_trace_cli.py:
```diff
@@ -0,0 +1,291 @@
+        - return: list of mme_id and enodebd_id in a list
+        """
+        hex_pattern = "0X.*"
+        pattern_mme_enb_ue_id = ".*MME_UE_S1AP_ID = (\w+)\\b eNB_UE_S1AP_ID = (\w+)\\b"
```
```suggestion
        pattern_mme_enb_ue_id = r".*MME_UE_S1AP_ID = (\w+)\b eNB_UE_S1AP_ID = (\w+)\b"
```

ardzoht commented on lte/gateway/python/scripts/user_trace_cli.py:
```diff
@@ -0,0 +1,291 @@
+                ue_id["mme_ue_id"] = int(ue_id["mme_ue_id"], 16)
+                ue_id["enb_ue_id"] = int(ue_id["enb_ue_id"], 16)
+
+            if len(ue_id) > 0 and ue_id not in ue_id_list:
```
```suggestion
            if ue_id and ue_id not in ue_id_list:
```

ardzoht commented on lte/gateway/python/scripts/user_trace_cli.py:
```diff
@@ -0,0 +1,291 @@
+
+        for v in ue_id_list:
+            mme, enodeb = v.items()
+            if enodeb[1] is not "None":
```
```suggestion
            if enodeb[1] != "None":
```

ardzoht reviewed (changes requested):
left some comments, I suggest running https://docs.magmacore.org/docs/next/lte/dev_unit_testing#format-agw to format the python file

wallyrb commented:
> left some comments, I suggest running https://docs.magmacore.org/docs/next/lte/dev_unit_testing#format-agw to format the python file

thanks @ardzoht for reviewing. I addressed the comments, run the precommit format script and fix remaining lint errors(with the exception of "WPS306 Found class without a base class: UserTrace" not sure if I can clean that one)

Feel free to add any other suggestions. Thanks!

ardzoht commented on lte/gateway/python/scripts/user_trace_cli.py:
```diff
@@ -0,0 +1,302 @@
+        lines = self.lines
+        imsi = self.args.imsi
+        for line in lines:
+            ue_id = {}
```
```suggestion
            ue_id_dict= {}
```
for clarity through the code

ardzoht commented on lte/gateway/python/scripts/user_trace_cli.py:
```diff
@@ -0,0 +1,302 @@
+            mme_id = int(match[0], 16)
+            if mme_id not in mme_id_list:
+                mme_id_list.append(mme_id)
+                ue_id_list.append({"mme_ue_id": mme_id, "enb_ue_id": "None"})
```
no need to append `enb_ue_id` here

ardzoht commented on lte/gateway/python/scripts/user_trace_cli.py:
```diff
@@ -0,0 +1,302 @@
+        print('-' * 25)
+
+        for v in ue_id_list:
+            mme, enodeb = v.items()
```
if we don't append `enb_ue_id` on line 161, we can update the check to:
```suggestion
            mme_ue_id = v.get('mme_id')
            enodeb_ue_id = v.get('enb_ue_id')
            if not enodeb_ue_id:
                print...
            else:
               print...
            
```

wallyrb replied:
great suggestion! thank you, I have updated with the comments

ardzoht commented on lte/gateway/python/scripts/user_trace_cli.py:
```diff
@@ -0,0 +1,302 @@
+        lines = self.lines
+        self.get_ue_id_pattern(args)
+        for line in lines:
+            match = re.findall(self.parttern_str, line, re.IGNORECASE)
```
```suggestion
            match = re.findall(self.pattern_str, line, re.IGNORECASE)
```

ardzoht commented on lte/gateway/python/scripts/user_trace_cli.py:
```diff
@@ -0,0 +1,302 @@
+            "MME_UE_S1AP_ID ?=? 0x0*" + format(ue_id, 'x') + r"\b",
+            r"mme_ue_s1ap_id = \(" + str(ue_id) + r"\)",
+        ]
+        self.parttern_str = "|".join(pattern_list)
```
```suggestion
        self.pattern_str = "|".join(pattern_list)
```

ardzoht commented on lte/gateway/python/scripts/user_trace_cli.py:
```diff
@@ -0,0 +1,302 @@
+                    print('*' * 100)
+                    break
+
+    def get_ue_id_pattern(self, args):
```
```suggestion
    def set_ue_id_pattern(self, args):
```

ardzoht commented on lte/gateway/python/scripts/user_trace_cli.py:
```diff
@@ -0,0 +1,302 @@
+            session_trace: Filtered logs for a given imsi and mme id
+        """
+
+        errors_list = [
```
move this on top of the file and define it as constant
```suggestion
        SESSIOND_ERRORS_LIST = [
```

wallyrb replied:
thanks for the suggestion!

wallyrb commented:
thanks a lot for the suggestions @ardzoht . I have updated the PR

ardzoht reviewed (commented):
Thanks for the changes Wally, it looks good overall to me, the last part we need to do is to add the script to the `lte/gateway/python/setup.py` under scripts list. Once we have that I will approve the PR

wallyrb commented:
> Thanks for the changes Wally, it looks good overall to me, the last part we need to do is to add the script to the `lte/gateway/python/setup.py` under scripts list. Once we have that I will approve the PR

thanks @ardzoht will add the script on the list and will add also some documentation for this cli.

wallyrb commented:
- Added the script under `lte/gateway/python/setup.py`
- Set the mme log default path in the script to /var/log/mme.log
- Add documentation for the script(attached picture in the test plan)
- Some minor changes in description of parameters.

@ardzoht let me know if you have any additional comments. Thanks!

electronjoe reviewed (approved):
`/docs/` CODEOWNER LGTM

ardzoht reviewed (approved):
lgtm

themarwhal reviewed (approved):
This looks great! I'm noticing the GHA jobs are not being triggered on this PR. (Is this the same issue we're seeing with @sebathomas ? cc @quentinDERORY , @electronjoe )

sebathomas commented:
Github support finally resolved the CI issue over the weekend and explained what happened. Github Actions were disabled entirely for my account because my fork of magma/magma was incorrectly flagged for "mining and reaching tor relay nodes". Unfortunately, they could not explain which part of the workflow was flagged.

If you have the same issue, you should see a red box with an error at the top of the "Actions" page of any repository. The only way to get it fixed is to contact Github support.

wallyrb commented:
rebasing and fixing markdown lint doc errors that were failing.
Note:  Below file was failing the markdown lint check, fixing to pass this check.
docs/readmes/nms/organizations.md

ardzoht commented on lte/gateway/python/scripts/user_trace_cli.py:
```diff
@@ -0,0 +1,303 @@
+from pathlib import Path
+
+
+class UserTrace:
```
```suggestion
class UserTrace(object):
```
this will fix remaining pep8 linting error

wallyrb replied:
Updated, thanks for the suggestion @ardzoht !
